### PR TITLE
feat(metrics): support for direct metric publishing

### PR DESCRIPTION
**Pull request by dreynaud**
... without going through Spring events first

### CODE REVIEW

**Comment by dreynaud** [comment]
I still need to test it but I wanted to get some feedback about a few things

File: keiko-core/build.gradle [hunk 1]
```diff
@@ -5,5 +5,7 @@ dependencies {
   api "com.fasterxml.jackson.core:jackson-databind"
   api "org.springframework:spring-context"
 
+  implementation "com.netflix.spectator:spectator-api"
```
Comment (dreynaud) [inline keiko-core/build.gradle]: keiko-core now depends on spectator, is that a problem?

File: keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    _lastState.set(queue.readState())
+  }
+
+  /**
+   * Count of messages pushed to the queue.
+   */
+  private val MessagePushed.counter: Counter
+    get() = when (payload) {
+      // TODO: what to do about ApplicationAware?
+//      is ApplicationAware -> registry.counter("queue.pushed.messages", "application", (payload as ApplicationAware).application)
```
Comment (dreynaud) [inline keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt]: I had a question about this, it seems ApplicationAware is an orca notion and not a keiko notion. Everything else is very orca agnostic, but I'm not sure what to do with this tag. Are we ok to drop it?

File: keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      is MessageRescheduled -> event.counter.increment()
+      is MessageNotFound -> event.counter.increment()
+    }
+  }
+
+  val lastState: QueueState
+    get() = _lastState.get()
+  private val _lastState = AtomicReference<QueueState>(QueueState(0, 0, 0))
+
+  @Scheduled(fixedDelayString = "\${queue.depth.metric.frequency:30000}")
```
Comment (dreynaud) [inline keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt]: ok having an `@Scheduled` directly here in keiko-core? (migrated from orca)

File: keiko-spring/src/main/kotlin/com/netflix/spinnaker/config/QueueConfiguration.kt [hunk 4]
```diff
@@ -95,11 +98,14 @@ class QueueConfiguration {
...
+  @ConditionalOnMissingBean(EventPublisher::class)
+  fun queueEventPublisher() = NoopEventPublisher()
+
+  @Bean
+  @ConditionalOnProperty("queue.metrics.enabled", havingValue = "true", matchIfMissing = true)
+  fun queueMetricsPublisher(
+    registry: Registry,
+    clock: Clock,
+    queue: MonitorableQueue
+  ) = QueueMetricsPublisher(registry, clock, queue)
```
Comment (dreynaud) [inline keiko-spring/src/main/kotlin/com/netflix/spinnaker/config/QueueConfiguration.kt]: that's really the only change to the environment, we swap out the `ApplicationEventPublisher` for a `QueueMetricsPublisher` and bypass Spring events for direct metrics. I thought this was pretty neat.

File: keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt (hunk 2 above)
Comment (asher) [reply keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt]: I think this is fine to drop from keiko. Could do something like proxy the Queue class in orca and inspect the message attributes on queue.push in orca if it turns out the metric is useful.

File: keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt (hunk 3 above)
Comment (asher) [reply keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt]: yeah, already used in `QueueProcessor` in keiko-core actually.

File: keiko-core/build.gradle (hunk 1 above)
Comment (robfletcher) [reply keiko-core/build.gradle]: Is Spectator open source?

File: keiko-spring/src/main/kotlin/com/netflix/spinnaker/config/QueueConfiguration.kt (hunk 4 above)
Comment (robfletcher) [reply keiko-spring/src/main/kotlin/com/netflix/spinnaker/config/QueueConfiguration.kt]: Yep, this is how I was hoping you'd do it.

File: keiko-core/build.gradle (hunk 1 above)
Comment (dreynaud) [reply keiko-core/build.gradle]: yes, it's also not specific to Atlas. I'm just mindful of new dependencies, especially in `*-core`

**Pull request merged**